src/arcaflow_plugin_sdk/schema.py

```diff
@@ -1,78 +1,3009 @@
+"""
+This module provides the tools to build a schema, either by hand, or automatically from type hints and annotations.
+
+Using
+=====
+
+For using this module please check the documentation located at
+https://arcalot.github.io/arcaflow/creating-plugins/python/
+
+Contributing
+============
+
+This module is structured by using region folding delimiters (e.g. # region SomeRegion). You can use an editor that
+supports these regions to collapse them together for easier access. The module has the following regions:
+
+Exceptions
+----------
+
+This region defines a number of exceptions that can be raised either while building a schema or while working with
+input and output.
+
+Annotations
+-----------
+
+Annotations can be applied to dataclass fields to convey extra information (e.g. minimum or maximum values). These
+annotations are used when building a schema. Note that, unlike tools like Pydantic we do not validate the dataclass
+fields in-situ. Instead, we only add metadata information which is verified when the data is serialized or unserialized.
+This late verification is done to avoid confusing error messages when using third party code that may be constructing
+the dataclasses with intermediate states that are temporarily invalid.
+
+Type aliases
+------------
+
+This section defines a number of type aliases for use in the schema itself.
+
+Schema
+------
+
+This section contains the data model for the Arcaflow schema definition itself. The schema is self-describing and
+can be serialized and unserialized with the same tools that you use to serialize or unserialize data that conforms to
+the schema. The plugin system uses these dataclasses to construct the SCOPE_SCHEMA and SCHEMA_SCHEMA variables, which
+you can use to serialize and unserialize a scope definition or an entire schema.
+
+Types
+-----
+
+Types are the implementations of a schema. They add serialization, validation, and unserialization to the schema itself.
+When working with schemas in practice you will want to work with types unless you don't need to unserialize data.
+
+Build
+-----
+
+This region holds the tools to automatically build a schema from annotations. Use build_object_schema to build a schema
+from a dataclass.
+
+Note that the built schema will be a type, not a schema, so it is possible to use these schemas for serializing and
+unserializing data.
+
+Schema schemas
+--------------
+
+This region holds two variables, SCOPE_SCHEMA and SCHEMA_SCHEMA. You can use these variables to work with schemas
```

Comment:
Just a suggestion: use rST syntax for variables: using \``SCOPE_SCHEMA\`` as example

Reply:
But can go into separate patch addressing documentation review feedback